### PR TITLE
Migration Attributes

### PR DESCRIPTION
**Pull request by ArtificialOwl**
### tasks

- [x] generate metadata from core+available apps (apps are temporary enabled during the process if needed)
- [x] download file from dl.nc.c
- [x] extract data and compare with already applied Migrations
- [x] emulate a generic entry in case a new Attribute is implemented in a future version
- [x] display data using the occ command
- [ ] making endpoint to display result on the webclient _(not done yet, but easily doable as code is now out of Command file)_
- [x] commenting and polishing the code
 
### new occ commands

- running `./occ migrations:generate-metadata` to create a JSON file 
- running `./occ migrations:preview <version>` to display migration that will be executed if an upgrade to _version_ is initiated

### details

When adding some attributes to a migration:
```
#[DropTable(table: 'old_table')]
#[CreateTable(table: 'new_table', description: 'Table is used to store things, but also to get more things', notes: ['this is a notice', 'and another one, if really needed'])]
#[AddColumn(table: 'my_table')]
#[AddColumn(table: 'my_table', name: 'another_field')]
#[AddColumn(table: 'other_table', name: 'last_one', type: ColumnType::DATE)]
#[AddIndex(table: 'my_table')]
#[AddIndex(table: 'my_table', type: IndexType::PRIMARY)]
#[DropColumn(table: 'other_table')]
#[DropColumn(table: 'other_table', name: 'old_column', description: 'field is not used anymore and replaced by \'last_one\'')]
#[DropIndex(table: 'other_table')]
#[ModifyColumn(table: 'other_table')]
#[ModifyColumn(table: 'other_table', name: 'this_field')]
#[ModifyColumn(table: 'other_table', name: 'this_field', type: ColumnType::BIGINT)]
class Version30000Date20240102030405 extends SimpleMigrationStep {
```

will display some details when running the preview occ command:

## steps to execute implemented new commands:

when using an install made from repository (with apps/testing/ available):

**generate metadata from current install:**
>     ./occ migrations:generate-metadata > /tmp/nc.metadata

**enable testing app:**
>     ./occ app:enable testing

**delete migration from db:**
>     DELETE FROM oc_migrations WHERE `oc_migrations`.`app` = 'testing' AND `oc_migrations`.`version` = '30000Date20240102030405'

**preview _future_ migrations:**
>     ./occ migrations:preview /tmp/nc.metadata     

_result:_
image

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) https://github.com/nextcloud/documentation/pull/12074
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/public/Migration/Attributes/MigrationWeight.php [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	/** @since 30.0.0 */
+	case LIGHT = 'light'; // migration should be light and quick
+	/** @since 30.0.0 */
+	case MEDIUM = 'medium'; // migration is estimated to require few minutes
+	/** @since 30.0.0 */
+	case SETUP_BASED = 'setup-based'; // depends on setup, migration might require some time
+	/** @since 30.0.0 */
+	case SIZE_RELATED = 'size-related'; // depends on setup, migration might require some time
+	/** @since 30.0.0 */
+	case HEAVY = 'heavy'; // migration is estimated to be heavy and time consuming
```
Comment (ChristophWurst) [inline lib/public/Migration/Attributes/MigrationWeight.php:21]: This is out of scope. We don't want to guess this. We don't know if adding a column is light or heavy.

**Comment by ChristophWurst** [comment]
> * `TableCreation`
> 
> * `TableDeletion`
> 
> * `FieldAdd`
> 
> * `FieldChange`
> 
> * `FieldRemove`
> 
> * `IndexAdd`
> 
> * `IndexRemove`

Ease the naming and avoid ambiguity by using standard sql terms

* CreateTable
* DeleteTable
* AddColumn
* ModifyColumn (consider more specific attributes for type change, rename, etc)
* DropColumn
* AddIndex
* DropIndex

File: lib/public/Migration/Attributes/MigrationWeight.php (hunk 1 above)
Comment (ArtificialOwl) [reply lib/public/Migration/Attributes/MigrationWeight.php:21]: _removed_

File: lib/public/Migration/Attributes/TableMigrationAttribute.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCP\Migration\Attributes;
+
+use JsonSerializable;
+
+class TableMigrationAttribute extends MigrationAttribute implements JsonSerializable {
+	public function __construct(
+		string $table = '',
```
Comment (ChristophWurst) [inline lib/public/Migration/Attributes/TableMigrationAttribute.php:15]:
```suggestion
		string $table,
```

table migrations without a name are impossible. the table name has to be mandatory

File: lib/public/Migration/Exceptions/AttributeException.php [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCP\Migration\Exceptions;
+
+use Exception;
+
+/**
+ * @since 30.0.0
+ */
+class AttributeException extends Exception {
```
Comment (ChristophWurst) [inline lib/public/Migration/Exceptions/AttributeException.php:16]: this is only used internally, move to oc

**Comment by ChristophWurst** [comment]
@AndyScherzinger this does not close https://github.com/nextcloud/server/issues/45943. It's the first of five changes

File: lib/public/Migration/Attributes/DeleteTable.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+/**
+ * SPDX-FileCopyrightText: 2024 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+namespace OCP\Migration\Attributes;
+
+use Attribute;
+
+#[Attribute(Attribute::IS_REPEATABLE | Attribute::TARGET_CLASS)]
+class DeleteTable extends MigrationAttribute {
```
Comment (ChristophWurst) [inline lib/public/Migration/Attributes/DeleteTable.php:14]:
```suggestion
class DropTable extends MigrationAttribute {
```

File: lib/public/Migration/Attributes/IndexType.php [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+/**
+ * @since 30.0.0
+ */
+enum IndexType : string {
+	/** @since 30.0.0 */
+	case PRIMARY = 'primary'; // migration is estimated to require few minutes
+	/** @since 30.0.0 */
+	case INDEX = 'index'; // depends on setup, migration might require some time
+	/** @since 30.0.0 */
+	case UNIQUE = 'unique'; // migration should be light and quick
```
Comment (ChristophWurst) [inline lib/public/Migration/Attributes/IndexType.php:20]: we don't want to estimate

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
* All new classes, their methods and properties require `@since 30.0.0`
* All new classes require unit test coverage

File: lib/public/Migration/Attributes/TableMigrationAttribute.php (hunk 2 above)
Comment (ArtificialOwl) [reply lib/public/Migration/Attributes/TableMigrationAttribute.php:15]: should be make it mandatory for any migration attribute ?

Comment (ChristophWurst) [reply lib/public/Migration/Attributes/TableMigrationAttribute.php:15]: Yes

File: lib/public/Migration/Attributes/MigrationAttribute.php [hunk 6]
```diff
@@ -14,6 +14,11 @@
  * @since 30.0.0
  */
 class MigrationAttribute implements JsonSerializable {
+	/**
+	 * @param string $table name of the database table
+	 * @param string $description description of the migration
+	 * @param array $notes notes about the migration
```
Comment (ChristophWurst) [inline lib/public/Migration/Attributes/MigrationAttribute.php:20]: @since is not inherited. You have to add it to *all* methods

Comment (ArtificialOwl) [reply lib/public/Migration/Attributes/MigrationAttribute.php:20]: oups!

**Comment by ArtificialOwl** [comment]
_psalm error not related: https://github.com/nextcloud/server/pull/46846_

**Comment by ArtificialOwl** [comment]
when using an install made from repository (with apps/testing/ available):

**generate metadata from current install:**
>     ./occ migrations:generate-metadata > /tmp/nc.metadata

**enable testing app:**
>     ./occ app:enable testing

**delete migration from db:**
>     DELETE FROM oc_migrations WHERE `oc_migrations`.`app` = 'testing' AND `oc_migrations`.`version` = '30000Date20240102030405'

**preview _future_ migrations:**
>     ./occ migrations:preview /tmp/nc.metadata     


_result:_
image

**Review by ChristophWurst — APPROVED** [review]
Attributes look good

**Review by sorbaugh — APPROVED** [review]

**Comment by ChristophWurst** [comment]
image

Developer docs were updated but admin docs are still missing

**Pull request merged**